### PR TITLE
Allow setting of the resource finder on a per resource basis

### PR DESCRIPTION
**Pull request by lgebhardt**
This gets us closer to supporting alternate resource finders. This should work with multiple resource finders as long as relationships do not span the different data stores.

I wanted to keep the resource finders as mix-ins to allow existing code to override the `records` and other methods. In order to support a default resource finder and allow setting this with a method in the resource class definition the mix-in needed to delay loading. This was accomplished with `method_missing` to load the resource finder if it isn't loaded.

### All Submissions:

- [ ] I've checked to ensure there aren't other open [Pull Requests](https://github.com/cerebris/jsonapi-resources/pulls) for the same update/change.
- [ ] I've submitted a [ticket](https://github.com/cerebris/jsonapi-resources/issues) for my issue if one did not already exist.
- [ ] My submission passes all tests. (Please run the full test suite locally to cut down on noise from travis failures.)
- [ ] I've used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message or the description.
- [ ] I've added/updated tests for this change.

### New Feature Submissions:

- [ ] I've submitted an issue that describes this feature, and received the go ahead from the maintainers.
- [ ] My submission includes new tests.
- [ ] My submission maintains compliance with [JSON:API](http://jsonapi.org/).

### Bug fixes and Changes to Core Features:

- [ ] I've included an explanation of what the changes do and why I'd like you to include them.
- [ ] I've provided test(s) that fails without the change.

### Reviewer Checklist:
- [ ] Maintains compliance with JSON:API
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: lib/jsonapi/resource.rb [hunk 1]
```diff
@@ -1071,6 +1018,33 @@ def register_relationship(name, relationship_object)
...
+        end
+        include _resource_finder
+        self._resource_finder_included = true
+      end
+
+      def method_missing(m, *args, &block)
+        if _resource_finder_included
+          super
+        else
+          # handle the case where a resource finder has not been include yet. This should only happen once per class.
```
Comment (scottgonzalez) [inline lib/jsonapi/resource.rb]: s/include/included/

s/handle/Handle/ since this is a full sentence.

File: lib/jsonapi/resource.rb [hunk 2]
```diff
@@ -1071,6 +1018,33 @@ def register_relationship(name, relationship_object)
 
       private
 
+      def _resource_finder=(resource_finder)
```
Comment (scottgonzalez) [inline lib/jsonapi/resource.rb]: Is this only so you can clear the value on inheritance?

File: lib/jsonapi/resource.rb [hunk 3]
```diff
@@ -421,75 +421,22 @@ def inherited(subclass)
...
-      def find_included_fragments(_source_rids, _relationship_name, _options = {})
-        # :nocov:
-        raise 'Abstract ResourceFinder method called. Ensure that a ResourceFinder has been set.'
-        # :nocov:
-      end
+        # Ensure that the parent's resource finder is included before inheriting from the parent is completed
+        if !_resource_finder_included && self != JSONAPI::Resource
+          include_resource_finder
+        end
+        subclass._resource_finder = nil
```
Comment (scottgonzalez) [inline lib/jsonapi/resource.rb]: Why is this needed? Shouldn't this already be `nil` at this point because the interpreter hasn't gotten past `class SubClass < ParentClass`? I didn't see anything that would indicate the resource finder is being copied from the parent to the child.

**Review by scottgonzalez — CHANGES_REQUESTED** [review]

File: lib/jsonapi/resource.rb (hunk 3 above)
Comment (lgebhardt) [reply lib/jsonapi/resource.rb]: Good catch. This, and the call you caught above, were left over from an earlier approach. I've removed them.

File: lib/jsonapi/resource.rb (hunk 2 above)
Comment (lgebhardt) [reply lib/jsonapi/resource.rb]: Removed - see below.

File: lib/jsonapi/resource.rb [hunk 4]
```diff
@@ -1070,6 +1016,28 @@ def register_relationship(name, relationship_object)
...
+        end
+        include _resource_finder
+        self._resource_finder_included = true
+      end
+
+      def method_missing(m, *args, &block)
+        if _resource_finder_included
+          super
+        else
+          # Handle the case where a resource finder has not been include yet. This should only happen once per class.
```
Comment (scottgonzalez) [inline lib/jsonapi/resource.rb]: You missed changing "include" to "included"

Comment (lgebhardt) [reply lib/jsonapi/resource.rb]: D'oh. Fixed

**Review by scottgonzalez — APPROVED** [review]
